SNT/NativeMemoryManager.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System.Buffers;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System;
+
+namespace Torch.SNT
+{
+    internal class NativeMemory<T> : MemoryManager<T>
+    {
+        private int refCount = 0;
+        private IntPtr memory;
+        private int length;
+
+        public NativeMemory(IntPtr memory, int length)
+        {
+            this.memory = memory;
+            this.length = length;
+        }
+
+        public unsafe NativeMemory(void* memory, int length)
+        {
+            this.memory = (IntPtr)memory;
+            this.length = length;
+        }
+
+        /// <summary>
+        /// The Tensor memory lifecycle is managed by Torch. Disposing the memory
+        /// from here will through a "double free or corruption" error.
+        /// <summary>
+        ~NativeMemory()
+        {
+            Dispose(false);
+        }
+
+        public static NativeMemory<T> Allocate(int length)
+        {
+            // typically this would call into a native method appropriate for the platform
+            // or the constructors above would be used to wrap the native pointer
+            IntPtr memory = Marshal.AllocHGlobal(Marshal.SizeOf<T>() * length);
+            return new NativeMemory<T>(memory, length);
+        }
+
+        public bool IsDisposed { get; private set; } = false;
+
+        public unsafe override Span<T> GetSpan() => new Span<T>((void*)memory, length);
+
+        protected bool IsRetained => refCount > 0;
+
+        public override MemoryHandle Pin(int elementIndex = 0)
```

Comment:
Doc comments on all public methods, please.

Reply:
Fixed

Reply:
FWIW this isn't a public type.